File: object.ts
import { Memory } from './memory';

export abstract class BoObject {
  abstract str(): string;
  compare(other: BoObject): number {
    if (this === other) return 0;
    else return -1;
  }
  clone(): BoObject {
    return this;
  }
}

export class Num extends BoObject {
  constructor(public value: number) {
    super();
  }
  str(): string {
    return this.value.toString();
  }
  compare(other: BoObject): number {
    if (other instanceof Num) {
      const result = this.value - other.value;
      if (result === 0) return 0;
      else if (result < 0) return -1;
      else return 1;
    }
    return -1;
  }
}

export class Str extends BoObject {
  constructor(public value: string) {
    super();
  }
  str(): string {
    return `"${this.value}"`;
  }
  compare(other: BoObject): number {
    if (other instanceof Str) {
      if (this.value === other.value) return 0;
      else if (this.value < other.value) return -1;
      else return 1; //this.value > other.value
    }
    return -1;
  }
  concat(other: BoObject): BoObject {
    if (other instanceof Str) return new Str(this.value + other.value);
    else return new Str(this.value + other.str());
  }
}

export class Nil extends BoObject {
  private static instance = new Nil();
  private constructor() {
    super();
  }
  str(): string {
    return 'nil';
  }
  compare(other: BoObject): number {
    if (other instanceof Nil) return 0;
    else return -1;
  }
  static getInstance(): Nil {
    return Nil.instance;
  }
}
export const NIL = Nil.getInstance();

export class Message extends BoObject {
  receiver: BoObject | undefined;
  slotName: string;
  args?: BoObject[];
  isTailCall: boolean;
  constructor(
    receiver: BoObject | undefined,
    slotName: string,
    args?: BoObject[]
  ) {
    super();
    this.receiver = receiver;
    this.isTailCall = false;
    if (slotName !== '.') {
      this.slotName = slotName;
      this.args = args;
    } else if (args?.length === 1 && args[0] instanceof Message) {
      this.slotName = args[0].slotName;
      this.args = args[0].args;
    } else {
      throw new Error(
        `ERROR new Message(${receiver?.str()},${slotName},${args
          ?.map((e) => e.str())
          .join()})`
      );
    }
  }
  str(): string {
    const receiverStr = this.receiver ? this.receiver.str() : '';
    let argsStr: string;
    if (this.args) {
      argsStr = '(' + this.args.map((e) => e.str()).join(', ') + ')';
    } else {
      argsStr = '';
    }
    if (this.slotName === '.' && this.args?.length === 1) {
      return receiverStr + '.' + this.args![0].str();
    } else {
      const slotName = this.slotName; //+ (this.isTailCall ? '$' : '');
      if (this.receiver === undefined) {
        return slotName + argsStr;
      } else {
        return receiverStr + '.' + slotName + argsStr;
      }
    }
  }
}

export class Fun extends BoObject {
  argList: string[];
  body: BoObject;
  createdEnv: Memory;
  constructor(args: BoObject[], env: Memory) {
    super();
    if (args.length === 0) {
      throw new Error('ERROR new Fun(no-argument)');
    }
    this.body = args[args.length - 1];
    this.argList = args.slice(0, args.length - 1).map((e) => {
      if (e instanceof Message && !e.receiver && !e.args) return e.slotName;
      else throw new Error(`ERROR new Fun() => type error. value=${e.str()}`);
    });
    this.createdEnv = env;
    this.findTailCall(this.body);
  }
  str(): string {
    const argStr =
      this.argList.length === 0 ? '' : this.argList.join(',') + ',';
    return `fun(${argStr}${this.body.str()})`;
  }
  findTailCall(node: BoObject): void {
    if (node instanceof Message && node.args) {
      if (node.slotName === 'if' || node.slotName === ';') {
        if (node.args[1]) this.findTailCall(node.args[1]);
        if (node.args[2]) this.findTailCall(node.args[2]);
      } else if (/^[a-zA-Z_]/.test(node.slotName)) {
        // user defined functions
        node.isTailCall = true;
      }
    }
  }
}

export class Macro extends BoObject {
  argList: string[];
  body: BoObject;
  constructor(args: BoObject[]) {
    super();
    if (args.length === 0) {
      throw new Error('ERROR new Macro(no-argument)');
    }
    this.body = args[args.length - 1];
    this.argList = args.slice(0, args.length - 1).map((e) => {
      if (e instanceof Message && !e.receiver && !e.args) return e.slotName;
      else throw new Error(`ERROR new Macro()  value=${e.str()}`);
    });
  }
  str(): string {
    const argStr =
      this.argList.length === 0 ? '' : this.argList.join(',') + ',';
    return `macro(${argStr}${this.body.str()})`;
  }
}

export class UserObject extends BoObject {
  memory: Memory;
  proto?: UserObject;
  constructor(slot: Memory, proto?: UserObject) {
    super();
    this.memory = slot;
    this.proto = proto;
  }
  compare(other: BoObject): number {
    if (this === other) return 0;
    else return -1;
  }
  str(): string {
    const s = [...this.memory.slots.entries()]
      .map(([k, v]) => {
        return k + ':' + v.str();
      })
      .join(',');
    return '{' + s + '}';
  }
  clone(): UserObject {
    return new UserObject(this.memory.subMemory(), this);
  }
  define(name: string, value: BoObject): BoObject | null {
    return this.memory.define(name, value);
  }
  update(name: string, value: BoObject): BoObject | null {
    return this.memory.update(name, value);
  }
  assignToObject(name: string, value: BoObject): BoObject {
    return this.memory.defineForce(name, value);
  }
  get(name: string): BoObject | undefined {
    return this.memory.get(name);
  }
}

export class TailCallNotification {
  constructor(
    public _this: UserObject | undefined,
    public fun: Fun,
    public args: BoObject[],
    public callerEnv: Memory
  ) {}
}

File: go-interpreter/object.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

// BoObject is the base interface for all objects in BinOp language.
type BoObject interface {
	Str() string
	Compare(other BoObject) int
	Clone() BoObject
}

// --- Num ---

type Num struct {
	Value float64
}

func NewNum(v float64) *Num {
	return &Num{Value: v}
}

func (n *Num) Str() string {
	return strconv.FormatFloat(n.Value, 'f', -1, 64)
}

func (n *Num) Compare(other BoObject) int {
	if o, ok := other.(*Num); ok {
		result := n.Value - o.Value
		if result == 0 {
			return 0
		} else if result < 0 {
			return -1
		} else {
			return 1
		}
	}
	return -1
}

func (n *Num) Clone() BoObject {
	return n
}

// --- Str ---

type Str struct {
	Value string
}

func NewStr(v string) *Str {
	return &Str{Value: v}
}

func (s *Str) Str() string {
	return fmt.Sprintf(`"%s"`, s.Value)
}

func (s *Str) Compare(other BoObject) int {
	if o, ok := other.(*Str); ok {
		if s.Value == o.Value {
			return 0
		} else if s.Value < o.Value {
			return -1
		} else {
			return 1
		}
	}
	return -1
}

func (s *Str) Clone() BoObject {
	return s
}

func (s *Str) Concat(other BoObject) BoObject {
	if o, ok := other.(*Str); ok {
		return NewStr(s.Value + o.Value)
	}
	return NewStr(s.Value + other.Str())
}

// --- Nil ---

type NilObj struct{}

var NIL = &NilObj{}

func (n *NilObj) Str() string {
	return "nil"
}

func (n *NilObj) Compare(other BoObject) int {
	if _, ok := other.(*NilObj); ok {
		return 0
	}
	return -1
}

func (n *NilObj) Clone() BoObject {
	return n
}

// --- Message ---

type Message struct {
	Receiver BoObject   // nil means no receiver
	SlotName string
	Args     []BoObject // nil means no args (undefined), empty slice means ()
	HasArgs  bool       // true if args were provided (distinguishes undefined vs [])
}

func NewMessage(receiver BoObject, slotName string, args []BoObject) *Message {
	hasArgs := args != nil

	if slotName != "." {
		return &Message{
			Receiver: receiver,
			SlotName: slotName,
			Args:     args,
			HasArgs:  hasArgs,
		}
	}

	// slotName === "."
	if args != nil && len(args) == 1 {
		if m, ok := args[0].(*Message); ok {
			return &Message{
				Receiver: receiver,
				SlotName: m.SlotName,
				Args:     m.Args,
				HasArgs:  m.HasArgs,
			}
		}
	}

	var argsStrs []string
	if args != nil {
		for _, e := range args {
			argsStrs = append(argsStrs, e.Str())
		}
	}
	recStr := ""
	if receiver != nil {
		recStr = receiver.Str()
	}
	panic(fmt.Sprintf("ERROR new Message(%s,%s,%s)", recStr, slotName, strings.Join(argsStrs, ",")))
}

func (m *Message) Str() string {
	receiverStr := ""
	if m.Receiver != nil {
		receiverStr = m.Receiver.Str()
	}

	argsStr := ""
	if m.HasArgs {
		parts := make([]string, len(m.Args))
		for i, e := range m.Args {
			parts[i] = e.Str()
		}
		argsStr = "(" + strings.Join(parts, ", ") + ")"
	}

	if m.SlotName == "." && m.Args != nil && len(m.Args) == 1 {
		return receiverStr + "." + m.Args[0].Str()
	} else if m.Receiver == nil {
		return m.SlotName + argsStr
	} else {
		return receiverStr + "." + m.SlotName + argsStr
	}
}

func (m *Message) Compare(other BoObject) int {
	if m == other {
		return 0
	}
	return -1
}

func (m *Message) Clone() BoObject {
	return m
}

// --- Fun ---

type Fun struct {
	ArgList    []string
	Body       BoObject
	CreatedEnv *Memory
}

func NewFun(args []BoObject, env *Memory) *Fun {
	if len(args) == 0 {
		panic("ERROR new Fun(no-argument)")
	}
	body := args[len(args)-1]
	argList := make([]string, 0, len(args)-1)
	for _, e := range args[:len(args)-1] {
		if m, ok := e.(*Message); ok && m.Receiver == nil && !m.HasArgs {
			argList = append(argList, m.SlotName)
		} else {
			panic(fmt.Sprintf("ERROR new Fun() => type error. value=%s", e.Str()))
		}
	}
	return &Fun{ArgList: argList, Body: body, CreatedEnv: env}
}

func (f *Fun) Str() string {
	argStr := ""
	if len(f.ArgList) > 0 {
		argStr = strings.Join(f.ArgList, ",") + ","
	}
	return fmt.Sprintf("fun(%s%s)", argStr, f.Body.Str())
}

func (f *Fun) Compare(other BoObject) int {
	if f == other {
		return 0
	}
	return -1
}

func (f *Fun) Clone() BoObject {
	return f
}

// --- Macro ---

type Macro struct {
	ArgList []string
	Body    BoObject
}

func NewMacro(args []BoObject) *Macro {
	if len(args) == 0 {
		panic("ERROR new Macro(no-argument)")
	}
	body := args[len(args)-1]
	argList := make([]string, 0, len(args)-1)
	for _, e := range args[:len(args)-1] {
		if m, ok := e.(*Message); ok && m.Receiver == nil && !m.HasArgs {
			argList = append(argList, m.SlotName)
		} else {
			panic(fmt.Sprintf("ERROR new Macro()  value=%s", e.Str()))
		}
	}
	return &Macro{ArgList: argList, Body: body}
}

func (mc *Macro) Str() string {
	argStr := ""
	if len(mc.ArgList) > 0 {
		argStr = strings.Join(mc.ArgList, ",") + ","
	}
	return fmt.Sprintf("macro(%s%s)", argStr, mc.Body.Str())
}

func (mc *Macro) Compare(other BoObject) int {
	if mc == other {
		return 0
	}
	return -1
}

func (mc *Macro) Clone() BoObject {
	return mc
}

// --- UserObject ---

type UserObject struct {
	Mem   *Memory
	Proto *UserObject
}

func NewUserObject(slot *Memory, proto *UserObject) *UserObject {
	return &UserObject{Mem: slot, Proto: proto}
}

func (u *UserObject) Compare(other BoObject) int {
	if u == other {
		return 0
	}
	return -1
}

func (u *UserObject) Str() string {
	parts := []string{}
	for _, k := range u.Mem.Order {
		if v, ok := u.Mem.Slots[k]; ok {
			parts = append(parts, k+":"+v.Str())
		}
	}
	return "{" + strings.Join(parts, ",") + "}"
}

func (u *UserObject) Clone() BoObject {
	return NewUserObject(u.Mem.SubMemory(), u)
}

func (u *UserObject) Define(name string, value BoObject) BoObject {
	return u.Mem.Define(name, value)
}

func (u *UserObject) Update(name string, value BoObject) BoObject {
	return u.Mem.Update(name, value)
}

func (u *UserObject) AssignToObject(name string, value BoObject) BoObject {
	return u.Mem.DefineForce(name, value)
}

func (u *UserObject) Get(name string) BoObject {
	return u.Mem.Get(name)
}
